stor/dx.py

```diff
@@ -385,129 +1067,161 @@ def exists(self):
+                                       .format(self.project))
+
+        return proj_dict['id']
+
     @cached_property
     def canonical_resource(self):
         """Returns the dx file-ID of the uniquely matched filename
 
         Raises:
             DuplicateError: if filename is not unique
             NotFoundError: if resource is not found on DX platform
+            ValueError: if path is folder path
         """
         if not self.resource:
             return None
-        if self._is_folder():
-            raise ValueError('DXPath ({}) ending in folders cannot be canonicalized'.format(self))
+        if utils.has_trailing_slash(self):
+            raise ValueError('Invalid operation ({method}) on folder path ({path})'
+                             .format(path=self, method=sys._getframe(2).f_code.co_name))
         objects = [{
             'name': self.name,
             'folder': self.resource.parent,
-            'project': self.canonical_project
+            'project': self.canonical_project,
+            'batchsize': 2
         }]
-        results = dxpy.resolve_data_objects(objects=objects)[0]
+        with _propagate_dx_exceptions():
+            results = dxpy.resolve_data_objects(objects=objects)[0]
         if len(results) > 1:
-            raise DuplicateError('The virtual resource is not unique on DNAnexus')
+            raise DuplicateError('Multiple objects found at path ({}). '
+                                 'Try using a canonical ID instead'.format(self))
         elif len(results) == 1:
             return results[0]['id']
         else:
-            raise NotFoundError('The virtual resource does not exist on DNAnexus')
+            raise stor_exceptions.NotFoundError(
+                'No data object was found for the given path ({}) on DNAnexus'.format(self))
 
     @property
     def canonical_path(self):
-        """Returns the unique file that matches the given path"""
-        return DXCanonicalPath(self.drive + self.canonical_project +
-                               ':') / (self.canonical_resource or '')
+        """Returns the unique file or project that matches the given path"""
+        return DXCanonicalPath('{drive}{proj_id}:/{resource}'.format(
+            drive=self.drive, proj_id=self.canonical_project,
+            resource=(self.canonical_resource or '')))
 
 
 class DXCanonicalPath(DXPath):
-    """Class Handler for DXPath of form 'dx://project-{ID}:/file-{ID}' or 'dx://project-{ID}:'"""
+    """Class Handler for DXPath in canonicalized form:
+    'dx://project-{dxID}:/file-{dxID}' or 'dx://project-{dxID}:'
+    """
 
     @property
     def virtual_project(self):
+        """Returns the virtual name of the project associated with the DXCanonicalPath"""
         return self.virtual_path.project
 
     @property
     def virtual_resource(self):
+        """Returns the human readable path of the resource associated with the DXCanonicalPath"""
```

Comment:
```suggestion
        """The virtual (human-readable) path of the resource associated with this path"""
```

btw - is there a DXPath class that has these but raises `NotImplemented`?

Reply:
no, but adding them to DXPath now.